Jetpack: Use image-cdn package

haqadn opened the pull request:
## Proposed changes:
We recently published a new package [image-cdn](https://github.com/Automattic/jetpack/tree/trunk/projects/packages/image-cdn) to be able to share the image cdn functionality with boost. This PR replaces the photon functionality which was within Jetpack and uses the package instead.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
pc9hqz-1FV-p2

## Does this pull request change what data or activity we track or use?
No

## Testing instructions:
* Test if image-cdn(Jetpack > Settings > Performance > Speed up image load times) works without boost
* Test if image-cdn works with boost.
* Test if the following modules still work as they had photon based dependency: 
![image](https://user-images.githubusercontent.com/3737780/231429656-2ed59a41-5d5f-485b-8c0f-133e5aebab5d.png)

haqadn committed:
Replace photon in Jetpack with the image-cdn package

haqadn committed:
Changelog

haqadn committed:
changelog

haqadn committed:
Fix docblock

haqadn committed:
Add shim functions for photon with deprecation call

haqadn committed:
Fix dependency

jeherve committed:
Undo unnecessary lock file changes

thingalon commented on projects/plugins/jetpack/functions.photon.php:
```diff
@@ -3,339 +3,79 @@
  *
  * @see https://developer.wordpress.com/docs/photon/
  *
+ * @deprecated 9.5 Use Automattic\Jetpack\Image_CDN\Image_CDN_Core::cdn_url instead.
```
9.5? Aren't we up to Jetpack 12?

haqadn replied:
Thank you for noticing. I thought I checked the current version, and not sure I ended up with 9.5. It appears the upcoming version is 12.1. I have update it.

thingalon reviewed (dismissed):
The changes look good to me, and work well - but the deprecation version doesn't look right to me. Aren't we up to Jetpack 12?

I also see failing tests.

thingalon commented:
I would also like to get a review from @Automattic/jetpack-crew on this.

haqadn committed:
Update deprecated since version

haqadn committed:
Remove non-existing test file reference

jeherve commented on projects/plugins/jetpack/class.photon.php:
```diff
@@ -1,1389 +0,0 @@
-/**
- * Class Jetpack_Photon
- */
-class Jetpack_Photon {
```
I do not think we can remove the class right away. There are quite a few plugins and themes in the directory that rely on that class today (see a few [here](https://wpdirectory.net/search/01GY7GYT6YAY4YK7EK91DACR0F), and we wouldn't want to create fatals for all of them. Some of them do check if `Jetpack_Photon` exists before to use it, but not all.

I would recommend keeping the class around, but adding deprecation warnings for all public methods while pointing them to the new class.

haqadn replied:
I have added backward compatibility for the file. Please take another look. 🙂

haqadn committed:
Keep `Jetpack_Photon` as a depricated class while forwarding method calls to Image_CDN.

haqadn committed:
Merge branch 'trunk' into update/use-image-cdn-package

haqadn committed:
[not verified] Remove deprication notice as a test

haqadn committed:
[not verified] Revert "[not verified] Remove deprication notice as a test"

This reverts commit 65a7934a1d57dcc659982840d6d68c3da2ef6939.

haqadn committed:
Merge branch 'trunk' into update/use-image-cdn-package

jeherve reviewed (commented):
> Test if image-cdn(Jetpack > Settings > Performance > Speed up image load times) works without boost

Is this working for you? In my tests on 2 sites, I cannot get any Photonized URLs when using the Jetpack plugin and its Photon module.

Tests are also failing at the moment:
https://github.com/Automattic/jetpack/actions/runs/4794114532/jobs/8535227272?pr=30050

haqadn committed:
Fix compatibility

haqadn committed:
Require AMP support class in test that depends on it

haqadn commented:
> > Test if image-cdn(Jetpack > Settings > Performance > Speed up image load times) works without boost
> 
> Is this working for you? In my tests on 2 sites, I cannot get any Photonized URLs when using the Jetpack plugin and its Photon module.
> 
> Tests are also failing at the moment: https://github.com/Automattic/jetpack/actions/runs/4794114532/jobs/8535227272?pr=30050

Nice catch. It stopped working after restoring `Jetpack_Photon` class as the compatibility file was checking existence of the class to make the module inactive to avoid conflict between old jetpack and Boost with Image CDN. I have fixed it now.

I also fixed an issue in e2e tests.

jeherve commented on projects/packages/image-cdn/src/compatibility/photon.php:
```diff
@@ -12,19 +12,30 @@
+	 *
+	 * When this is happening, the package is likely being loaded from a different plugin, e.g. Boost.
+	 */
+	if ( class_exists( 'Jetpack' ) && defined( 'JETPACK__VERSION' ) && ( version_compare( JETPACK__VERSION, '12.0', '<=' ) ) ) {
```
Since we're past the code freeze for 12.1, I think this will need to be updated to 12.1.

jeherve replied:
2f8a02d46c0e44572b9a87572c90bf04a9978190

jeherve commented on projects/plugins/jetpack/functions.photon.php:
```diff
@@ -3,339 +3,79 @@
-	}
-
-	return $args;
+	_deprecated_function( __FUNCTION__, 'jetpack-12.1', 'Automattic\Jetpack\Image_CDN\Image_CDN_Core::parse_wpcom_query_args' );
```
Those will need to be updated to 12.2. You can use `$$next-version$$` here to not have to worry about versions in your PRs. During release, `$$next-version$$` will be automatically replaced by the right number.

```suggestion
	_deprecated_function( __FUNCTION__, 'jetpack-$$next-version$$', 'Automattic\Jetpack\Image_CDN\Image_CDN_Core::parse_wpcom_query_args' );
```

jeherve commented on projects/plugins/jetpack/functions.photon.php:
```diff
@@ -3,339 +3,79 @@
-	}
-
-	return preg_replace( '#^([a-z:]+)?//#i', $scheme_slashes, $url );
+	_deprecated_function( __FUNCTION__, 'jetpack-12.1', 'Automattic\Jetpack\Image_CDN\Image_CDN_Core::cdn_url_scheme' );
```
```suggestion
	_deprecated_function( __FUNCTION__, 'jetpack-$$next-version$$', 'Automattic\Jetpack\Image_CDN\Image_CDN_Core::cdn_url_scheme' );
```

jeherve commented on projects/plugins/jetpack/functions.photon.php:
```diff
@@ -3,339 +3,79 @@
-	}
-
-	return $skip;
+	_deprecated_function( __FUNCTION__, 'jetpack-12.1', 'Automattic\Jetpack\Image_CDN\Image_CDN_Core::banned_domains' );
```
```suggestion
	_deprecated_function( __FUNCTION__, 'jetpack-$$next-version$$', 'Automattic\Jetpack\Image_CDN\Image_CDN_Core::banned_domains' );
```

jeherve commented on projects/plugins/jetpack/functions.photon.php:
```diff
@@ -3,339 +3,79 @@
 /**
  * Jetpack Photon - Support Text Widgets.
  *
+ * @deprecated 12.1 Use Automattic\Jetpack\Image_CDN\Image_CDN_Core::support_text_widgets instead.
```
```suggestion
 * @deprecated $$next-version$$ Use Automattic\Jetpack\Image_CDN\Image_CDN_Core::support_text_widgets instead.
```

jeherve commented on projects/plugins/jetpack/functions.photon.php:
```diff
@@ -3,339 +3,79 @@
-		return Jetpack_Photon::filter_the_content( $content );
-	}
-	return $content;
+	_deprecated_function( __FUNCTION__, 'jetpack-12.1', 'Automattic\Jetpack\Image_CDN\Image_CDN_Core::support_text_widgets' );
```
```suggestion
	_deprecated_function( __FUNCTION__, 'jetpack-$$next-version$$', 'Automattic\Jetpack\Image_CDN\Image_CDN_Core::support_text_widgets' );
```

jeherve commented on projects/plugins/jetpack/functions.photon.php:
```diff
@@ -3,339 +3,79 @@
 /**
  * Check to skip Photon for a known domain that shouldn't be Photonized.
  *
+ * @deprecated 12.1 Use Automattic\Jetpack\Image_CDN\Image_CDN_Core::banned_domains instead.
```
```suggestion
 * @deprecated $$next-version$$ Use Automattic\Jetpack\Image_CDN\Image_CDN_Core::banned_domains instead.
```

jeherve commented on projects/plugins/jetpack/functions.photon.php:
```diff
@@ -3,339 +3,79 @@
 /**
  * Sets the scheme for a URL
  *
+ * @deprecated 12.1 Use Automattic\Jetpack\Image_CDN\Image_CDN_Core::cdn_url_scheme instead.
```
```suggestion
 * @deprecated $$next-version$$ Use Automattic\Jetpack\Image_CDN\Image_CDN_Core::cdn_url_scheme instead.
```

jeherve commented on projects/plugins/jetpack/functions.photon.php:
```diff
@@ -3,339 +3,79 @@
 /**
  * Parses WP.com-hosted image args to replicate the crop.
  *
+ * @deprecated 12.1 Use Automattic\Jetpack\Image_CDN\Image_CDN_Core::parse_wpcom_query_args instead.
```
```suggestion
 * @deprecated $$next-version$$ Use Automattic\Jetpack\Image_CDN\Image_CDN_Core::parse_wpcom_query_args instead.
```

jeherve commented on projects/plugins/jetpack/functions.photon.php:
```diff
@@ -3,339 +3,79 @@
-	}
-
-	return jetpack_photon_url_scheme( $photon_url, $scheme );
+	_deprecated_function( __FUNCTION__, 'jetpack-12.1', 'Automattic\Jetpack\Image_CDN\Image_CDN_Core::cdn_url' );
```
```suggestion
	_deprecated_function( __FUNCTION__, 'jetpack-$$next-version$$', 'Automattic\Jetpack\Image_CDN\Image_CDN_Core::cdn_url' );
```

jeherve commented on projects/plugins/jetpack/functions.photon.php:
```diff
@@ -3,339 +3,79 @@
  *
  * @see https://developer.wordpress.com/docs/photon/
  *
+ * @deprecated 12.1 Use Automattic\Jetpack\Image_CDN\Image_CDN_Core::cdn_url instead.
```
```suggestion
 * @deprecated $$next-version$$ Use Automattic\Jetpack\Image_CDN\Image_CDN_Core::cdn_url instead.
```

jeherve commented on projects/packages/image-cdn/src/compatibility/photon.php:
```diff
@@ -12,19 +12,30 @@
 	remove_filter( 'jetpack_photon_url', 'jetpack_photon_url', 10, 3 );
 	remove_filter( 'jetpack_photon_pre_args', 'jetpack_photon_parse_wpcom_query_args', 10, 2 );
 	remove_filter( 'jetpack_photon_skip_for_url', 'jetpack_photon_banned_domains', 9, 2 );
 	remove_filter( 'widget_text', 'jetpack_photon_support_text_widgets' );
```
I'm a bit confused with this one. Wouldn't we want to only run this when the module is active? Right now, we end up using Photon for images in widgets even when the module is not active.

haqadn replied:
More context here: pdWQjU-gk-p2#comment-321

I am assuming you are referring to the filters themselves and not to `remove_filter` calls. The package adds those filters regardless of whether the module is active, just like Photon originally did.

It feels wrong to filter widget text even when the module isn't active. And I think it should be changed. But I don't want to change any functionality in this PR.

jeherve replied:
> It feels wrong to filter widget text even when the module isn't active.

Yes, that's what I was getting at.

> And I think it should be changed. But I don't want to change any functionality in this PR.

With latest stable, we do not filter widget text when the module isn't active. Your PR changes that. I don't think it should.

jeherve replied:
Addressed in 7ee9a7efc0e7c98efe270d53a640e5a805db22cc

jeherve commented on projects/plugins/jetpack/class.photon.php:
```diff
@@ -3,1387 +3,51 @@
 /**
  * Class Jetpack_Photon
+ *
+ * @deprecated 12.1 Use Automattic\Jetpack\Image_CDN\Image_CDN instead.
```
```suggestion
 * @deprecated $$next-version$$ Use Automattic\Jetpack\Image_CDN\Image_CDN instead.
```

jeherve commented on projects/plugins/jetpack/class.photon.php:
```diff
@@ -3,1387 +3,51 @@
  * Class for photon functionality.
  *
  * @package automattic/jetpack
+ * @deprecated 12.1 Use Automattic\Jetpack\Image_CDN\Image_CDN instead.
```
```suggestion
 * @deprecated $$next-version$$ Use Automattic\Jetpack\Image_CDN\Image_CDN instead.
```

jeherve commented on projects/plugins/jetpack/class.photon.php:
```diff
@@ -3,1387 +3,51 @@
-			$tags['og:image'] = $images;
+	public function __call( $name, $arguments ) {
+		if ( method_exists( Image_CDN::class, $name ) ) {
+			_deprecated_function( __CLASS__ . '::' . esc_html( $name ), 'jetpack-12.1', 'Automattic\Jetpack\Image_CDN\Image_CDN::' . esc_html( $name ) );
```
```suggestion
			_deprecated_function( __CLASS__ . '::' . esc_html( $name ), 'jetpack-$$next-version$$', 'Automattic\Jetpack\Image_CDN\Image_CDN::' . esc_html( $name ) );
```

jeherve commented on projects/plugins/jetpack/class.photon.php:
```diff
@@ -3,1387 +3,51 @@
-			return;
+	public static function __callStatic( $name, $arguments ) {
+		if ( method_exists( Image_CDN::class, $name ) ) {
+			_deprecated_function( __CLASS__ . '::' . esc_html( $name ), 'jetpack-12.1', 'Automattic\Jetpack\Image_CDN\Image_CDN::' . esc_html( $name ) );
```
```suggestion
			_deprecated_function( __CLASS__ . '::' . esc_html( $name ), 'jetpack-$$next-version$$', 'Automattic\Jetpack\Image_CDN\Image_CDN::' . esc_html( $name ) );
```

jeherve commented on projects/plugins/jetpack/tests/php/modules/shortcodes/test-class.archives.php:
```diff
@@ -1,5 +1,6 @@
 <?php
 
+require_once JETPACK__PLUGIN_DIR . '3rd-party/class.jetpack-amp-support.php';
```
Were the tests failing before? This seems unrelated to the PR so that's a bit confusing. When I comment this out, the tests still work for me locally.

haqadn replied:
This file was included in photon tests, and that change happened in trunk while the PR was open. This is because a couple of unit tests crossover between amp and photon. I marked those as skipped and will be moving those back to jetpack in a separate PR.

Because we have moved all unit test files from jetpack to the package, it started to fail jetpack tests after I merged trunk into this PR.

jeherve reviewed (changes requested):
This tests well for me overall. Nice work with this, that's quite the project :)

I left some comments below, little things compared to that project.

**I should note that I did not test the PR on WordPress.com.**

jeherve committed:
Update deprecation version to placeholder

See https://github.com/Automattic/jetpack/pull/30050/files#r1179438193

jeherve commented:
I updated all the references to the plugin version for you so you do not have to go through all of those yourself. :)

jeherve committed:
[not verified] Merge remote-tracking branch 'origin/trunk' into update/use-image-cdn-package

jeherve committed:
Update JP version when the image CDN will be included

jeherve commented on projects/plugins/jetpack/_inc/lib/class.jetpack-photon-image.php:
Since we're deleting that file, we'll need to remove this reference:

https://github.com/Automattic/jetpack/blob/6886717a6d906983b938c8aa6f0295ab36b72463/projects/plugins/jetpack/_inc/lib/admin-pages/class-jetpack-redux-state-helper.php#L286

jeherve reviewed (changes requested):
Should we take this opportunity to delete the whole `projects/plugins/jetpack/modules/photon` directory? It only has the `photon.js` file in it now, and that file is now loaded from the package.

We would then be able to remove this reference as well:
https://github.com/Automattic/jetpack/blob/cb9c765e5c776107cf1df3f79417b59e4aafd0da/projects/plugins/jetpack/tools/webpack.config.js#L80

This brings up the question of minifying the `photon.js` file that is now served from the package. Since it is loaded on every page of the site, it is important for it to be minified. 
I would recommend adding a basic webpack config file to the new package, a bit like it's done in `projects/packages/lazy-images/webpack.config.js`, and minify the `photon.js` file from the package.
You can then edit the file enqueuing logic in `projects/packages/image-cdn/src/class-image-cdn.php` so it enqueues minified files when necessary. You can check how it's done here:
https://github.com/Automattic/jetpack/blob/3a3c4adc1c0913bac224ddb832713bd6101fe1ab/projects/packages/lazy-images/src/lazy-images.php#L514-L523

Since the Image CDN package already requires the Assets package, this should be possible without too much trouble.

haqadn committed:
[not verified] Use image-cdn on widgets only when the module is active

haqadn committed:
Remove deleted file reference

haqadn committed:
Remove photon js from webpack builds

haqadn committed:
Minify image-cdn file

haqadn committed:
Merge branch 'trunk' into update/use-image-cdn-package

haqadn committed:
Update lock files

haqadn committed:
changelog

haqadn committed:
Update composer lock

haqadn committed:
Update jetpack composer lock

jeherve commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -1211,16 +1213,18 @@ public function action_wp_enqueue_scripts() {
-			true
+
+		Assets::register_script(
+			'jetpack-image-cdn',
```
Maybe we can keep calling it `jetpack-photon`, just in case anyone has done some customizing to the enqueue on their site based on that script handle?

haqadn replied:
Will address in a separate PR

jeherve commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -1211,16 +1213,18 @@ public function action_wp_enqueue_scripts() {
+			)
 		);
+
+		Assets::enqueue_script( 'jetpack-image-cdn' );
```
You can get rid of the `Assets::enqueue_script` method if you pass `'enqueue' => true` as a parameter in the array above.

jeherve replied:
Done in 01fa4bab9740decf8228b1ea02bb2ac550b56054

jeherve committed:
Ship new dist directory in built package

jeherve committed:
Simplify enqueue

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into update/use-image-cdn-package

jeherve commented on projects/plugins/jetpack/functions.photon.php:
```diff
@@ -3,339 +3,79 @@
-	}
-
-	return jetpack_photon_url_scheme( $photon_url, $scheme );
+	_deprecated_function( __FUNCTION__, 'jetpack-$$next-version$$', 'Automattic\Jetpack\Image_CDN\Image_CDN_Core::cdn_url' );
```
Coming back to this, I have mixed feelings about this.

- On one hand, we want plugin and theme authors to start using the new package instead of the legacy feature.
- On the other hand, there are many themes and plugins using the function today. Some of those themes have more than 30,000 installs. This deprecation notice, while only visible in the logs on most sites, will generate some support questions for Jetpack and for those theme authors. It will even generate notices on WordPress.com where we use `jetpack_photon_url` in many places.

With that in mind, I'm thinking we may want to keep this specific function around without a deprecation notice, as an exception.

What do you think?

haqadn replied:
I agree, and this has concerned me as well. Since we have the depreciation marked in the doc block, it should be alright. I'll address it in a separate PR.

jeherve reviewed (approved):
This tests well for me. I only have 2 comments left, but they're not a blocker, I think it could be addressed in a follow-up PR.

Nice work on that monster of a PR! 👏
